File: infra/monitor.py
#!/usr/bin/env python3
"""
Bevakningsskript för politiker-webapp.
Hämtar nya motioner från riksdagen + lokala nyheter från SVT,
lagrar i D1-tabellen monitored_items.
Körs av systemd-timer (monitor.timer) dagligen.
"""
import urllib.request, urllib.error, xml.etree.ElementTree as ET
import json, os, sys, hashlib, logging
from datetime import datetime, timezone

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s monitor %(levelname)s %(message)s",
    handlers=[logging.StreamHandler(sys.stdout)],
)
log = logging.getLogger()

ENV_FILE = os.path.expanduser("~/.appdata/.config/.env")
CF_ACCOUNT_ID = "b74f8c0c6a92f3006483840cf27372fd"
CF_DB_ID = "e9ecf94f-fa71-4004-a5b8-f9317eb4d4e9"
UA = "Mozilla/5.0 (compatible; politikerkontakt-monitor/1.0)"

# SVT lokalnyheter per SVT-region → area_name i DB
SVT_REGIONS = {
    "blekinge":         "Blekinge",
    "dalarna":          "Dalarna",
    "gavleborg":        "Gävleborgs län",
    "halland":          "Region Halland",
    "jamtland":         "Jämtlands län",
    "jonkoping":        "Jönköpings län",
    "norrbotten":       "Norrbottens län",
    "skane":            "Region Skåne",
    "smaland":          "Kronobergs län",
    "stockholm":        "Stockholms län",
    "sormland":         "Södermanlands län",
    "uppsala":          "Uppsala län",
    "varmland":         "Värmlands län",
    "vast":             "Västra Götalandsregionen",
    "vasterbotten":     "Västerbottens län",
    "vasternorrland":   "Västernorrlands län",
    "vastmanland":      "Västmanlands län",
    "orebro":           "Örebro läns landsting",
    "ost":              "Region Östergötland",
}

def load_env():
    env = {}
    if os.path.exists(ENV_FILE):
        with open(ENV_FILE) as f:
            for line in f:
                line = line.strip()
                if "=" in line and not line.startswith("#"):
                    k, _, v = line.partition("=")
                    env[k.strip()] = v.strip().strip('"').strip("'")
    return env

def fetch(url, timeout=15):
    req = urllib.request.Request(url, headers={"User-Agent": UA})
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return r.read()

def item_id(url):
    return hashlib.sha1(url.encode()).hexdigest()[:20]

def parse_rss(content):
    root = ET.fromstring(content)
    items = []
    for item in root.findall(".//item"):
        title = item.findtext("title", "").strip()
        link = item.findtext("link", "").strip()
        pub = item.findtext("pubDate", "").strip()
        desc = item.findtext("description", "").strip()
        if title and link:
            items.append({"title": title, "url": link, "published": pub, "summary": desc[:500]})
    return items

def d1_query(sql, cf_token, params=None):
    url = f"https://api.cloudflare.com/client/v4/accounts/{CF_ACCOUNT_ID}/d1/database/{CF_DB_ID}/query"
    body = json.dumps({"sql": sql, "params": params or []}).encode()
    req = urllib.request.Request(
        url, data=body,
        headers={"Authorization": f"Bearer {cf_token}", "Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=30) as r:
        data = json.loads(r.read())
    if not data.get("success"):
        raise RuntimeError(f"D1 error: {data.get('errors')}")
    return data["result"][0]

def ensure_table(cf_token):
    d1_query("""
        CREATE TABLE IF NOT EXISTS monitored_items (
            id TEXT PRIMARY KEY,
            source TEXT NOT NULL,
            item_type TEXT NOT NULL,
            title TEXT NOT NULL,
            url TEXT NOT NULL,
            area_name TEXT,
            area_type TEXT,
            summary TEXT,
            published_at INTEGER,
            created_at INTEGER NOT NULL,
            letter_queued INTEGER NOT NULL DEFAULT 0
        )
    """, cf_token)

def upsert_items(items, cf_token):
    inserted = 0
    now = int(datetime.now(timezone.utc).timestamp() * 1000)
    BATCH = 50
    for start in range(0, len(items), BATCH):
        batch = items[start:start + BATCH]
        placeholders = ",".join("(?,?,?,?,?,?,?,?,?,?)" for _ in batch)
        params = []
        for item in batch:
            params += [
                item["id"], item["source"], item["item_type"],
                item["title"], item["url"],
                item.get("area_name"), item.get("area_type"),
                item.get("summary", ""),
                item.get("published_at"), now,
            ]
        try:
            result = d1_query(
                f"INSERT OR IGNORE INTO monitored_items (id, source, item_type, title, url, area_name, area_type, summary, published_at, created_at) VALUES {placeholders}",
                cf_token, params,
            )
            inserted += result["meta"].get("changes", 0)
        except Exception as e:
            log.warning("Batch %d misslyckades: %s", start, e)
    return inserted

def fetch_riksdagen_motioner():
    sources = [
        ("mot", "motion"),
        ("prop", "proposition"),
        ("bet", "betankande"),
    ]
    items = []
    for doktyp, item_type in sources:
        try:
            content = fetch(
                f"https://data.riksdagen.se/dokumentlista/?doktyp={doktyp}&utformat=rss&sort=datum&sortorder=desc&sz=50"
            )
            for entry in parse_rss(content):
                items.append({
                    "id": item_id(entry["url"]),
                    "source": "riksdagen",
                    "item_type": item_type,
                    "title": entry["title"],
                    "url": entry["url"],
                    "area_name": None,
                    "area_type": "riksdag",
                    "summary": entry["summary"],
                    "published_at": None,
                })
            log.info("Riksdagen %s: %d poster", doktyp, len(items))
        except Exception as e:
            log.warning("Riksdagen %s misslyckades: %s", doktyp, e)
    return items

def fetch_svt_local():
    items = []
    for slug, area_name in SVT_REGIONS.items():
        try:
            content = fetch(f"https://www.svt.se/nyheter/lokalt/{slug}/rss.xml")
            for entry in parse_rss(content):
                items.append({
                    "id": item_id(entry["url"]),
                    "source": "svt_lokal",
                    "item_type": "news",
                    "title": entry["title"],
                    "url": entry["url"],
                    "area_name": area_name,
                    "area_type": "region",
                    "summary": entry["summary"],
                    "published_at": None,
                })
        except Exception as e:
            log.warning("SVT %s misslyckades: %s", slug, e)
    log.info("SVT lokalt: %d poster totalt", len(items))
    return items

def main():
    env = load_env()
    cf_token = env.get("CLOUDFLARE_API_TOKEN_POLITIKER")
    if not cf_token:
        log.error("CLOUDFLARE_API_TOKEN_POLITIKER saknas i %s", ENV_FILE)
        sys.exit(1)

    ensure_table(cf_token)

    all_items = []
    all_items += fetch_riksdagen_motioner()
    all_items += fetch_svt_local()

    inserted = upsert_items(all_items, cf_token)
    log.info("Klart: %d nya poster av %d insamlade", inserted, len(all_items))

if __name__ == "__main__":
    main()

File: campaign/src/monitor.ts
import type { Env } from "./index";

const SVT_REGIONS: Record<string, string> = {
  blekinge: "Blekinge", dalarna: "Dalarna", gavleborg: "Gävleborg",
  halland: "Halland", helsingborg: "Helsingborg", jonkoping: "Jönköping",
  norrbotten: "Norrbotten", skane: "Skåne", smaland: "Småland",
  stockholm: "Stockholm", soder: "Södertälje", ost: "Region Östergötland",
  vast: "Väst", vasterbotten: "Västerbotten", vasternorrland: "Västernorrland",
  vastmanland: "Västmanland", orebro: "Örebro", uppsala: "Uppsala",
  sormland: "Södermanland",
};

const NATIONAL_SOURCES = [
  { url: "https://www.svt.se/nyheter/rss.xml",                                source: "svt_national",  area: "Sverige" },
  { url: "https://www.svt.se/nyheter/utrikes/rss.xml",                        source: "svt_utrikes",   area: "Utrikes" },
  { url: "https://rss.aftonbladet.se/rss2/small/pages/sections/senastenytt/", source: "aftonbladet",   area: "Sverige" },
  { url: "https://feeds.expressen.se/nyheter/",                               source: "expressen",     area: "Sverige" },
];

const RIKSDAG_TYPES = [
  { type: "mot", label: "motion" },
  { type: "prop", label: "proposition" },
  { type: "bet", label: "betankande" },
];

interface RssItem { title: string; url: string; summary: string }

function parseRss(xml: string): RssItem[] {
  const items: RssItem[] = [];
  for (const m of xml.matchAll(/<item>([\s\S]*?)<\/item>/g)) {
    const item = m[1];
    const title = (item.match(/<title><!\[CDATA\[([\s\S]*?)\]\]>/) ?? item.match(/<title>([\s\S]*?)<\/title>/))?.[1]?.trim() ?? "";
    const url   = (item.match(/<link>([\s\S]*?)<\/link>/) ?? item.match(/<guid[^>]*>(https?:\/\/[\s\S]*?)<\/guid>/))?.[1]?.trim() ?? "";
    const desc  = (item.match(/<description><!\[CDATA\[([\s\S]*?)\]\]>/) ?? item.match(/<description>([\s\S]*?)<\/description>/))?.[1] ?? "";
    const summary = desc.replace(/<[^>]*>/g, "").replace(/</g, "").trim().slice(0, 500);
    if (title && url) items.push({ title, url, summary });
  }
  return items;
}

async function itemId(url: string): Promise<string> {
  const hash = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(url));
  return Array.from(new Uint8Array(hash)).map(b => b.toString(16).padStart(2, "0")).join("").slice(0, 16);
}

async function fetchRss(url: string): Promise<string> {
  const resp = await fetch(url, { headers: { "User-Agent": "politiker-webapp-monitor/1.0" }, signal: AbortSignal.timeout(15000) });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
  return resp.text();
}

interface DbItem {
  id: string; source: string; item_type: string; title: string;
  url: string; area_name: string; area_type: string; summary: string;
}

async function batchUpsert(env: Env, items: DbItem[]): Promise<number> {
  let inserted = 0;
  for (let i = 0; i < items.length; i += 50) {
    const batch = items.slice(i, i + 50);
    const stmts = batch.map(it =>
      env.DB.prepare(
        "INSERT OR IGNORE INTO monitored_items (id,source,item_type,title,url,area_name,area_type,summary,letter_queued,created_at) VALUES (?,?,?,?,?,?,?,?,0,?)"
      ).bind(it.id, it.source, it.item_type, it.title, it.url, it.area_name, it.area_type, it.summary, Date.now())
    );
    const results = await env.DB.batch(stmts);
    inserted += results.reduce((s, r) => s + (r.meta.changes ?? 0), 0);
  }
  return inserted;
}

export async function runMonitor(env: Env): Promise<void> {
  const items: DbItem[] = [];

  // Riksdagen RSS
  for (const { type, label } of RIKSDAG_TYPES) {
    try {
      const xml = await fetchRss(`https://www.riksdagen.se/sv/sok/?avd=dokument&doktyp=${type}&utformat=rss&sidnr=1`);
      for (const entry of parseRss(xml)) {
        items.push({
          id: await itemId(entry.url), source: "riksdagen", item_type: label,
          title: entry.title, url: entry.url, area_name: "Riksdagen",
          area_type: "riksdag", summary: entry.summary,
        });
      }
    } catch (e) { console.error(`monitor: fel vid riksdagen/${type}:`, e); }
  }

  // SVT regionalt
  for (const [slug, areaName] of Object.entries(SVT_REGIONS)) {
    try {
      const xml = await fetchRss(`https://www.svt.se/nyheter/lokalt/${slug}/rss.xml`);
      for (const entry of parseRss(xml)) {
        items.push({
          id: await itemId(entry.url), source: `svt_${slug}`, item_type: "news",
          title: entry.title, url: entry.url, area_name: areaName,
          area_type: "region", summary: entry.summary,
        });
      }
    } catch (e) { console.error(`monitor: fel vid svt_${slug}:`, e); }
  }

  // Nationella källor
  for (const s of NATIONAL_SOURCES) {
    try {
      const xml = await fetchRss(s.url);
      for (const entry of parseRss(xml)) {
        items.push({
          id: await itemId(entry.url), source: s.source, item_type: "news",
          title: entry.title, url: entry.url, area_name: s.area,
          area_type: "riksdag", summary: entry.summary,
        });
      }
    } catch (e) { console.error(`monitor: fel vid ${s.source}:`, e); }
  }

  const inserted = await batchUpsert(env, items);
  console.log(`monitor: ${items.length} poster hämtade, ${inserted} nya`);
}
